### PR TITLE
Event Rec Model

### PR DESCRIPTION
**Pull request by Manj0t**
## Summary
Implements the Two-Tower event recommendation model end-to-end: training pipeline, inference script, and the Convex schema/query changes needed to support it. Also rewrites `updateUserPreferences.py` to produce the 3-slice `[attended | interested | blocked]` user feature vector that the UserTower consumes.

---
## Why is this change necessary?
The existing `updateUserPreferences.py` only captured attended-event signal, which isn't enough for a real recommender — we need interested and blocked signals too, each weighted differently, so the model can learn both positive and negative preference. On the scoring side, main had no actual recommender: user weights were written to Convex but nothing consumed them. This PR adds the Two-Tower model (UserTower + EventTower, 64-dim embeddings, BPR loss) along with the offline training pipeline and online inference path that reads features from Convex and ranks events per user.

Closes #

---
## Changes

**Convex schema (`schema.ts`)**
- Added `interactionType` field to `usersToEvents` as `"attended" | "interested" | "blocked"` — previously the table had no way to distinguish interaction types
- Added `by_userId_interactionType` compound index for efficient per-type lookups

**Convex queries (`data_ml/eventRec.ts`)**
- `getByUserId` now filters to `attended` only (preserves prior behavior for callers that assumed attended-only semantics)
- Added `getUserTagWeights({ userIDs })` — batch fetch of stored tag weight vectors, used by `recommendEvent.py` at inference
- Added `getInteractionsByUserId({ userId })` — returns all interactions so the Python side can split by `interactionType` when rebuilding user features

**Seed data (`eventSeedsStatic.ts`)**
- Added 8 new seed events (Chinatown Night Market, UNLV Career Fair, Streetwear Market, Desert Rap Cypher, Downtown Art Walk, Sunset Chill Fest, Neon Birthday Bash, UNLV Anime Club) to give the rec engine a more diverse event pool during testing

**`updateUserPreferences.py` (rewrite)**
- Now builds a `(3 * num_tags,)` user feature vector: `[attended_weights | interested_weights | blocked_weights]` — this is the input format the UserTower expects
- Per-interaction-type row weights: attended=1.0, interested=0.5, blocked=2.0 (blocked gets a stronger signal since it's an explicit negative)
- Tuned hyperparameters: `BETA 0.15 → 0.10`, `TAU 1.0 → 1.25`
- Cold-start priors from `userPreferredTags` are now clipped into attended_weights with `np.clip(att + prior * (1 - att), 0, 1)` instead of being added raw
- Refactored into smaller functions: `init_tags`, `event_multihot`, `build_matrix`, `build_weights`, `get_interaction_ids`, `build_user_feature_vector`

**Two-Tower model (new)**
- `twoTowerModel.py` — `UserTower` (3·num_tags → 128 → 128 → 64) and `EventTower` (num_tags+4 → 128 → 128 → 64), both L2-normalized outputs
- `twoTowerTrainer.py` — BPR loss, Adam with weight decay 1e-5, cosine annealing LR schedule (lr 3e-4 → 1e-5 over training)
- `trainEventRec.py` — training loop with train/test split, best-checkpoint restore by eval loss, graceful `KeyboardInterrupt` handling that still saves, partial state_dict load for shape-compatible resume
- `eventRecDataset.py` — `EventRecDataset` + `get_data_loader(mode="offline"|"online")`. Offline loads `.npy` files, online pulls live from Convex
- `generate_training_data.py` — generates synthetic `user_features.npy`, `event_features.npy`, `triplets.npy`, `tag_index.json` for offline training
- `recommendEvent.py` — inference entry point. Pulls user weights + event features from Convex, runs both towers, scores via `(u @ e.T + 1) / 2`, hard-masks blocked events to -1, returns top-k
- `utils.py` — shared `get_client()` helper for Convex connection

---
## UML Diagram (optional)
```mermaid
graph TD
    A[usersToEvents<br/>+ interactionType] --> B[updateUserPreferences.py]
    C[userPreferredTags<br/>cold start] --> B
    B -->|3 * num_tags vector| D[userTagWeights table]

    D --> E[recommendEvent.py]
    F[eventTags + event fields] --> E
    E --> G[UserTower]
    E --> H[EventTower]
    G --> I[cosine scores<br/>blocked events masked]
    H --> I
    I --> J[top-k per user]

    K[generate_training_data.py] --> L[.npy files]
    L --> M[trainEventRec.py]
    M --> N[TwoTowerTrainer<br/>BPR + cosine LR]
    N --> O[curr_model.pt]
    O --> E
```

---
## Testing

1. Pull the branch and apply the schema: ```cd packages/backend && pnpm convex dev```
2. Reseed so the new events + interaction types are populated: ```pnpm convex seed:seed```
3. Rebuild user feature vectors: ```cd packages/data-ml/event-rec``` and ```uv run python data/updateUserPreferences.py```
4. Confirm it writes to `userTagWeights` for all users (check Convex dashboard) and that `len(weights) == 3 * num_tags`.
5. Run recommendations ```uv run python recommendEvent.py```
(Rest Optional, Trust me it works): 
6. Generate fake training data ```uv run python data/generate_training_data.py```
7. Retrain from scratch to confirm the training loop works: ```uv run python -m training.trainEventRec```

### CODE REVIEW

**Review by Ekinn7188 — APPROVED** [review]
Im glad you fixed the issues along the way. LGTM

**Review by yahirescobar-unlv — APPROVED** [review]
Incredible. Mr. FAANG just called, you're in. Dr. Quant also called, he asked when can you start? You're looking at a $500,000,000 base salary.
